### PR TITLE
Limelight 3 code working

### PR DESCRIPTION
**Pull request by stardust4089**
Look, i know this should be in its own class. But if it is, it breaks. So, its going in periodic. For now. For GPK. I promise I will make it its own class before OSF.

Includes:
- Updates to `DrivebaseSubsystem.java` to make it detect tags properly through ll3 network tables
- `LimelightHelpers.java` for limelight lib to achieve above goals

This is messy but it works. It does not crash the robot and gives a correct and precise pose in under 20 ms and over 70 fps. That's half as much latency and 3 times as many frames as photonvision on the 2+. For that reason, this is needed for competition, and should be merged into main.

### CODE REVIEW

File: src/main/java/frc/robot/Vision/Limelight.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    private final NetworkTableEntry ledMode;
+    private final NetworkTableEntry camMode;
+    private final NetworkTableEntry pipeline;
+    private final NetworkTableEntry stream;
+    private final NetworkTableEntry snapshot;
+
+    private final NetworkTableEntry tid;
+    private final NetworkTableEntry botpose;
+
+    private LimelightResults llresults = LimelightHelpers.getLatestResults("limelight");
```
Comment (mikemag) [inline src/main/java/frc/robot/Vision/Limelight.java:35]: The reason your code works in `DrivebaseSubsysem.periodic` and not in this Limelight subsystem is right here. You're parsing your limelight data just once, on this line, when you construct the subsystem, and never again. You have methods below which refer to this one blip of data and never re-read it.

If the drivebase, you're actually re-reading and parsing your JSON limelight data every time. Which is correct.

File: src/main/java/frc/robot/subsystems/Swerve/DrivebaseSubsystem.java [hunk 2]
```diff
@@ -184,15 +190,29 @@ public Rotation2d getGyroscopeRotation() {
     public void periodic() {
 
         swerveOdometry.update(getGyroscopeRotation(), getModulePositions());
-
+        System.out.println(ll.getRobotPose());
```
Comment (mikemag) [inline src/main/java/frc/robot/subsystems/Swerve/DrivebaseSubsystem.java:193]: Strike this. This is the only place you use `ll` as well, so strike it too and do not create your broken `Limelight` subsystem. You're printing the same pose here every time... it comes from the single read of the limelight data in my other comment.

File: src/main/java/frc/robot/subsystems/Swerve/DrivebaseSubsystem.java [hunk 3]
```diff
@@ -184,15 +190,29 @@ public Rotation2d getGyroscopeRotation() {
...
         if (result.isPresent()) {
             EstimatedRobotPose camPose = result.get();
             swerveOdometry.addVisionMeasurement(
                     camPose.estimatedPose.toPose2d(),
                     camPose.timestampSeconds);
-        }*/
+        }
+        LimelightResults llresults = LimelightHelpers.getLatestResults("limelight");
+
+        NetworkTable table = NetworkTableInstance.getDefault().getTable("limelight");
```
Comment (mikemag) [inline src/main/java/frc/robot/subsystems/Swerve/DrivebaseSubsystem.java:204]: It feels weird to read a ton of data from the Limelight as JSON, and materialize a complex structure object from it, only to re-read three more variables from the device directly. I do not know if your second reads are from cache and consistent with the first read, or if you have inconsistent values here. If it's working now, roll with it for GPK, but it's a question you surely want to understand.

File: src/main/java/frc/robot/subsystems/Swerve/DrivebaseSubsystem.java [hunk 4]
```diff
@@ -184,15 +190,29 @@ public Rotation2d getGyroscopeRotation() {
...
+        NetworkTableEntry tl = table.getEntry("tl");
+        NetworkTableEntry cl = table.getEntry("cl");
+        NetworkTableEntry tv = table.getEntry("tv");
+
+        double totalLatency = cl.getDouble(1) + tl.getDouble(1);
+        Pose2d please_work = llresults.targetingResults.getBotPose2d_wpiBlue();
```
Comment (mikemag) [inline src/main/java/frc/robot/subsystems/Swerve/DrivebaseSubsystem.java:210]: What's the difference with red vs blue here? Do you need to use the red value when you're on red alliance?

File: src/main/java/frc/robot/subsystems/Swerve/DrivebaseSubsystem.java [hunk 5]
```diff
@@ -184,15 +190,29 @@ public Rotation2d getGyroscopeRotation() {
...
+        double hasTarget = tv.getDouble(0);
+
+        double im_crying = Timer.getFPGATimestamp() - (totalLatency / 1000.0);
+        if (hasTarget == 1)
```
Comment (mikemag) [inline src/main/java/frc/robot/subsystems/Swerve/DrivebaseSubsystem.java:214]: Does this `1` mean "I only see a single april tag"? Would be good to make a note in code or in the PR/commit why that'd be the limit right now.

**Review by mikemag — APPROVED** [review]
I know you're short on time, so I'm hitting "approve".

Do look at the comments, though. And consider removing your one use of the broken `Limelight` subsystem. Also consider naming that correctly before you commit and have to rename the file later. 

I'd say the only thing I'm really worried about for you for comp is the comment on red vs. blue.

**Comment by stardust4089** [comment]
Closing this PR due to refactoring happening post-GPK

**Pull request closed without merging**